packages/router/src/util.ts

```diff
@@ -251,6 +251,34 @@ function flattenAll(children: ReactNode): ReactNode[] {
   })
 }
 
+/**
+ *
+ * @param {string} queryString
+ * @returns {Array<string | Record<string, any>>} A flatten array of search params
+ *
+ * useMatch hook options searchParams requires flatten array
+ *
+ * Examples:
+ *
+ *  parseSearch('?key1=val1&key2=val2')
+ *  => { key1: 'val1', key2: 'val2' }
+ *
+ * flattenSearchParams(parseSearch('?key1=val1&key2=val2'))
+ * => [ { key1: 'val1' }, { key2: 'val2' } ]
+ *
+ */
+function flattenSearchParams(
+  queryString: string
+): Array<string | Record<string, any>> {
```

Comment:
The return type is a little unfortunate here since it would still be valid for the code to return `[{ one: '1', two: '2'}]` because that's an array with one element, and that element is a record (with two keys). Ideally the return type would limit it to just be strings or objects with one single key. But I'm not entirely sure how to do that. Maybe you can use `SingleKey` from here https://stackoverflow.com/a/60807986/88106

Reply:
From **type** point of view, you are absolutely right, it can be `single` or `multiple` keys, but from the **logic** point of view, it is guaranteed to **always** return `singleKey`
```jsx
for (const [key, value] of Object.entries(parseSearch(queryString))) {
  searchParams.push({ [key]: value })
}
```

with that said, the question now is *do we need to restrict the type to single key?* I will suggest keep it as `Record<string, any>`, why? if you look into the logic inside `useMatch`, the logic is compatible with `multiple` keys
```js
if (options?.searchParams) {
    const locationParams = new URLSearchParams(location.search)
    const hasUnmatched = options.searchParams.some((param) => {
      if (typeof param === 'string') {
        return !locationParams.has(param)
      } else {
        return Object.keys(param).some(  // <---------------- Compatible with single/multiple keys
          (key) => param[key] != locationParams.get(key)
        )
      }
    })

    if (hasUnmatched) {
      return { match: false }
    }
  }
```

so consuming `useMatch` externally, they both work
```jsx
const matchInfo = useMatch(pathname, { searchParams: [ { page: 1, sort: 'asc' } ] })
// or
const matchInfo = useMatch(pathname, { searchParams: [ { page: 1 }, { sort: 'asc' } ] })
```
above all, it's good to keep `flattenSearchParams` return as `Record<string, any>` so the logic can be extended in the future to potentially return `multiple` since `useMatch` does supports it

Reply:
> From **type** point of view, you are absolutely right, it can be `single` or `multiple` keys, but from the **logic** point of view, it is guaranteed to **always** return `singleKey`

For me, one of the main benefits of TypeScript is that I can just look at the types of a function to know what parameters it takes, and what it returns. But right now, for this function, that would be misleading and I would also have to look at the implementation.

Reply:
> For me, one of the main benefits of TypeScript is that I can just look at the types of a function to know what parameters it takes, and what it returns. But right now, for this function, that would be misleading and I would also have to look at the implementation.

The same, I rely heavily on Typescript to tell me the ability of utility functions. I still don't understand why you think it is misleading but I am keen to know your thinking logic as writing readable code for everyone is the key to success instead of writing something only I understand. 😄 

So, let's say I am trying to consume `useMatch` and VSCode shows me `searchParams` is an array of `string | Record<string, any>`, my instinct is okay I can provide something like `{ searchParams:  [  {  page:  1, sort:  'asc'  }  ]  }` then I will start wondering why has to be an array, why not an object, then I realise okay I can also provide `string` and look at the example in the comment, now I know why it's an array, then I will start using it like this `{ searchParams: [ 'tab', { page: 1, sort: 'asc'  }  ]  }` and it works as expected. I might not aware I can use like this `{ searchParams: [ 'tab', { page: 1 }, { sort: 'asc'  }  ]  }` but at least I am not misled.

That's how I will work through and trying to understand this, would be good to understand yours so we can figure out what makes you confused 😃 

<img width="875" alt="Screen Shot 2021-08-07 at 12 53 42 pm" src="https://user-images.githubusercontent.com/3888780/128585727-b40a458f-2c89-4066-8c9a-4f921a1a1a29.png">

<img width="497" alt="Screen Shot 2021-08-07 at 1 08 01 pm" src="https://user-images.githubusercontent.com/3888780/128586015-4b8a65f4-240b-4806-9f8e-b0526cbd6fdb.png">

Reply:
I think the confusion here is only that I didn't realize you made it work with both options :)  It's all good 👍